File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/f99a12495b2728af4769f86cc0d7abf7.py
from collections import defaultdict
import sys
import random

MOD = 998244353

class ModInt:
    def __init__(self, x=0):
        self.x = x % MOD
        if self.x < 0:
            self.x += MOD

    def __add__(self, other):
        return ModInt(self.x + other.x)

    def __sub__(self, other):
        return ModInt(self.x - other.x)

    def __mul__(self, other):
        return ModInt(self.x * other.x)

    def __truediv__(self, other):
        return ModInt(self.x * pow(other.x, MOD - 2, MOD))

    def __pow__(self, n):
        return ModInt(pow(self.x, n, MOD))

    def __repr__(self):
        return str(self.x)

class NTT:
    def __init__(self):
        self.max_base = 0
        self.root = 2
        while self.root ** ((MOD - 1) // 2) == 1:
            self.root += 1
        self.dw = []
        self.idw = []

        tmp = MOD - 1
        while tmp % 2 == 0:
            tmp //= 2
            self.max_base += 1

        self.dw = [-self.root ** ((MOD - 1) // (1 << (i + 2))) for i in range(self.max_base)]
        self.idw = [ModInt(1) / self.dw[i] for i in range(self.max_base)]

    def ntt(self, a):
        n = len(a)
        for m in range(n, 1, -1):
            w = ModInt(1)
            for s in range(0, n, 2 * m):
                for i in range(s, s + m):
                    j = i + m
                    x = a[i]
                    y = a[j] * w
                    a[i] = x + y
                    a[j] = x - y
                w *= self.dw[self.max_base - (m // 2).bit_length()]

    def intt(self, a, f=True):
        n = len(a)
        for m in range(1, n, 2):
            w = ModInt(1)
            for s in range(0, n, 2 * m):
                for i in range(s, s + m):
                    j = i + m
                    x = a[i]
                    y = a[j]
                    a[i] = x + y
                    a[j] = (x - y) * w
                w *= self.idw[self.max_base - (m // 2).bit_length()]
        if f:
            inv_sz = ModInt(1) / n
            for i in range(n):
                a[i] *= inv_sz

    def multiply(self, a, b):
        need = len(a) + len(b) - 1
        nbase = 1
        while (1 << nbase) < need:
            nbase += 1
        sz = 1 << nbase
        a += [ModInt(0)] * (sz - len(a))
        b += [ModInt(0)] * (sz - len(b))
        self.ntt(a)
        self.ntt(b)
        inv_sz = ModInt(1) / sz
        for i in range(sz):
            a[i] *= b[i] * inv_sz
        self.intt(a, False)
        return a[:need]

ntt = NTT()

def main():
    n, k = map(int, input().split())
    h = list(map(int, input().split()))
    h.append(h[0])

    cnt = sum(1 for i in range(n) if h[i] == h[i + 1])
    d = n - cnt

    res = ModInt(k) ** cnt
    sum_f = ModInt(0)

    f = [ModInt(0)] * (2 * d + 1)
    f[0] = ModInt(1)
    f[1] = ModInt(k - 2)
    f[2] = ModInt(1)

    f = pow(f, d)
    for i in range(d + 1, 2 * d + 1):
        sum_f += f[i]

    print(res * sum_f)

if __name__ == "__main__":
    main()

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/f99a12495b2728af4769f86cc0d7abf7.java
import java.util.*;
import java.io.*;

class Main {
    static final int MOD = 998244353;
    static int n, k;
    static int[] h;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        k = sc.nextInt();
        h = new int[n];
        for (int i = 0; i < n; i++) {
            h[i] = sc.nextInt();
        }
        h = Arrays.copyOf(h, n + 1);
        h[n] = h[0];

        int cnt = 0;
        for (int i = 0; i < n; i++) {
            if (h[i] == h[i + 1]) {
                cnt++;
            }
        }
        int d = n - cnt;

        long res = pow(k, cnt);
        long sum = 0;

        long[] f = new long[2 * d + 1];
        f[0] = 1;
        f[1] = k - 2;
        f[2] = 1;
        f = pow(f, d);
        for (int i = d + 1; i <= 2 * d; i++) {
            sum += f[i];
        }

        System.out.println(res * sum % MOD);
    }

    static long pow(long a, long b) {
        long res = 1;
        while (b > 0) {
            if ((b & 1) == 1) {
                res = res * a % MOD;
            }
            a = a * a % MOD;
            b >>= 1;
        }
        return res;
    }

    static long[] pow(long[] a, long b) {
        int n = a.length;
        long[] res = new long[n];
        res[0] = 1;
        while (b > 0) {
            if ((b & 1) == 1) {
                res = multiply(res, a);
            }
            a = multiply(a, a);
            b >>= 1;
        }
        return res;
    }

    static long[] multiply(long[] a, long[] b) {
        int n = a.length;
        long[] res = new long[n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j <= i; j++) {
                res[i] += a[j] * b[i - j] % MOD;
                res[i] %= MOD;
            }
        }
        return res;
    }
}
